hail/src/main/scala/is/hail/cxx/PackEncoder.scala

```diff
@@ -69,6 +87,7 @@ object PackEncoder {
     case _: PFloat64 => s"$output_buf_ptr->write_double($value);"
     case _: PBinary => encodeBinary(tub, output_buf_ptr, value)
     case t2: PArray => encodeArray(tub, t2, output_buf_ptr, value)
+    case t2: PNDArray => encodeNDArray(tub, t2, output_buf_ptr, value)
```

Comment:
This pattern makes me think: why not have a generic write IR node? The code emitted for NDArray is not special at all. However, what about multiple encoding strategies for the same data type? Something feels off but can't tell if it's too flexible or not flexible enough.

Reply:
You might want to just have your IR node take the CodecSpec here, I think. Is that what you mean by multiple encoding strategies? Eventually the physical layout will be in an EncodedType or something.

As for a general write IR node---that seems very reasonable. Our current thing for writing partitions is that we write each row followed by a byte (`1` or `0` denoting whether or not there are further rows). If we make that the default "encoding" for stream types, then what you've written seems like it might generalize to be able to write partitions, although there's a little bit of an additional question there on how to handle the path it gets written to.